### PR TITLE
Change `THPStorage::cdata` to be a `MaybeOwned<Storage>`, add unpack func

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96801

Part of #91395

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

### CODE REVIEW

File: torch/csrc/Storage.h [hunk 1]
```diff
@@ -5,17 +5,114 @@
 
 #define THPStorageStr "torch.UntypedStorage"
 
+namespace c10 {
+
+template <>
+struct MaybeOwnedTraits<c10::Storage> {
```
Comment (kurtamohler) [inline torch/csrc/Storage.h:11]: NOTE: I imagine these OwnedTraits defs should be moved to `c10/core/Storage.h`, but I have them here for now to reduce build time if I need to modify them

**Comment by ezyang** [comment]
do you want this reviewed now or later

**Comment by kurtamohler** [comment]
@ezyang, I think it's ready for a review

File: torch/csrc/Storage.cpp [hunk 2]
```diff
@@ -196,36 +194,34 @@ static PyObject* THPStorage_pynew(
...
-    uint8_t value = storage_get(
-        at::unsafeStorageFromTH(self->cdata, /*retain=*/true), nindex);
+    uint8_t value = storage_get(storage, nindex);
     return THPByteUtils_newReal(value);
     /* Slice index */
   } else if (PySlice_Check(index)) {
     // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
     Py_ssize_t start, stop, slicelength, step;
-    int64_t len = self->cdata->nbytes() / sizeof(uint8_t);
+    int64_t len = storage.nbytes();
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/Storage.cpp:224]: oh the claim is sizeof(uint8_t) must be one. okey dokey

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**